aten/src/ATen/FunctionalTensorImplBase.cpp

```diff
@@ -0,0 +1,184 @@
+
+namespace at {
+
+FunctionalTensorImplBase::FunctionalTensorImplBase(c10::DispatchKeySet keyset, caffe2::TypeMeta dtype, c10::Device device)
+  : TensorImpl(
+      keyset.add(DispatchKey::Functionalize),
+      dtype,
+      device
+    ) {}
+
+FunctionalTensorImplBase::FunctionalTensorImplBase(caffe2::TypeMeta dtype, c10::Device device)
+  : TensorImpl(
+      c10::DispatchKeySet(DispatchKey::Functionalize),
+      dtype,
+      device
+    ) {}
+
+void FunctionalTensorImplBase::mark_as_alias() {
+  is_alias_tensor_ = true;
+}
+
+void FunctionalTensorImplBase::sync_() {
+  if (is_up_to_date()) {
+    return;
+  }
+  // Apply all updates on alias_
+  alias_->SyncUpdateOperations();
+  // Reapply views to Get the viewed tensor from updated base in alias_
+  auto t = alias_->base();
+  for (auto& view_meta: view_metas_) {
+    switch (view_meta.view_type) {
+      case ViewMeta::Type::view:
+          t = t.view(view_meta.size);
+          break;
+      case ViewMeta::Type::noOp:
+          break;
+      default:
+          TORCH_CHECK(false, "Tried to run the functionalization pass on an unsupported view: ", view_meta.view_type);
+    }
+  }
+  // We want the new tensor to have separate memory from the alias.
+  // Note that the clone required for functorch, but probably unnecessary for backends
+  // like vulkan and xla, because their view operators already create fresh tensors.
+  t = t.clone();
+  // This call to replace_() doesn't go through the dispatcher.
+  // It's a virtual method implemented directly on TensorImpl subclasses.
+  replace_(t.unsafeGetTensorImpl());
+  generation_ = alias_->generation();
+}
+
+bool FunctionalTensorImplBase::is_up_to_date() const {
+  if (alias_) {
+    return generation_ == alias_->generation();
+  }
+  return true;
+}
+
+void FunctionalTensorImplBase::maybe_add_update(const Tensor& updated_val) {
+  // If the mutated tensor wasn't a view, we don't need to do anything
+  if (is_view()) {
+    alias_->add_update(updated_val.clone(), view_metas_);
+    // TODO: add this back see if anything breaks?
+    //generation_ = alias_->generation(); // self is fully up to date at this point
+  }
+}
+
+void FunctionalTensorImplBase::set_view_meta(const Tensor& other, at::ViewMeta meta) {
+    auto other_impl = dynamic_cast<at::FunctionalTensorImplBase*>(other.unsafeGetTensorImpl());
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(other_impl != nullptr);
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(alias_ == nullptr);
+    if (other_impl->alias_ != nullptr) {
+        // if the other tensor is already a view, copy its ViewMeta vector and push the current one.
+        std::vector<at::ViewMeta> metas = other_impl->view_metas_;  // copy
+        metas.push_back(meta);
+        //alias_ = other_impl->alias_; // refcount bump
+        alias_ = other_impl->alias_; // refcount bump
+        view_metas_ = metas;
+    } else {
+        std::shared_ptr<Alias> alias = std::make_shared<Alias>(const_cast<Tensor&>(other));
```

Comment:
Not sure what's going on with the const cast here. Just make Alias take a `const Tensor&`?

Reply:
I don't think I can make Alias take a `const Tensor&`, mostly because `other` and `alias->base_` are supposed to be distinct tensors (so `alias->base_` can't just hold a reference to `other`). That's because syncing the alias needs to apply updates to the base, which should happen independently from the actual tensors that know about the alias.

So I could make Alias take a `const Tensor` (non-reference), but to keep the `const` I would have to avoid re-assigning to `base_` when I apply updates to it (inside of `Alias::apply_update()`). I think I could do that with `replace_()` instead of re-assignment, but I'm also pretty sure that'd be slower - since it would require multiple trips through the dispatcher instead of a single assignment operator (and refcount bump).

lmk if you don't agree with that chain of reasoning